sw/snRuntime/src/dma.h

```diff
@@ -10,49 +10,15 @@ typedef uint32_t snrt_dma_txid_t;
 /// Initiate an asynchronous 1D DMA transfer with wide 64-bit pointers.
 inline snrt_dma_txid_t snrt_dma_start_1d_wideptr(uint64_t dst, uint64_t src,
                                                  size_t size) {
-    // Current DMA does not allow transfers with size == 0 (blocks)
-    // TODO(colluca) remove this check once new DMA is integrated
-    if (size > 0) {
-        register uint32_t reg_dst_low asm("a0") = dst >> 0;    // 10
-        register uint32_t reg_dst_high asm("a1") = dst >> 32;  // 11
-        register uint32_t reg_src_low asm("a2") = src >> 0;    // 12
-        register uint32_t reg_src_high asm("a3") = src >> 32;  // 13
-        register uint32_t reg_size asm("a4") = size;           // 14
-
-        // dmsrc a2, a3
-        asm volatile(
-            ".word (0b0000000 << 25) | \
-                (     (13) << 20) | \
-                (     (12) << 15) | \
-                (    0b000 << 12) | \
-                (0b0101011 <<  0)   \n" ::"r"(reg_src_high),
-            "r"(reg_src_low));
-
-        // dmdst a0, a1
-        asm volatile(
-            ".word (0b0000001 << 25) | \
-                (     (11) << 20) | \
-                (     (10) << 15) | \
-                (    0b000 << 12) | \
-                (0b0101011 <<  0)   \n" ::"r"(reg_dst_high),
-            "r"(reg_dst_low));
-
-        // dmcpyi a0, a4, 0b00
-        register uint32_t reg_txid asm("a0");  // 10
-        asm volatile(
-            ".word (0b0000010 << 25) | \
-                (  0b00000 << 20) | \
-                (     (14) << 15) | \
-                (    0b000 << 12) | \
-                (     (10) <<  7) | \
-                (0b0101011 <<  0)   \n"
-            : "=r"(reg_txid)
-            : "r"(reg_size));
-
-        return reg_txid;
-    } else {
-        return -1;
-    }
+    register uint32_t reg_txid;  // 10
+    asm volatile(
+        "dmsrc   %[sl], %[sh]\n"
+        "dmdst   %[dl], %[dh]\n"
+        "dmcpyi  %[id], %[sz], 0"
+        : [ id ] "=r"(reg_txid)
+        : [ sh ] "r"(src >> 32), [ sl ] "r"(src), [ dh ] "r"(dst >> 32),
+          [ dl ] "r"(dst), [ sz ] "r"(size));
```

Comment:
If it is not too much trouble, would it be okay to keep these in a form consumeable by non-pulp toolchains? Due to higher requirements from our toolchain we are using a normal Clang 18.1 to compile and run programs on Snitch instead of the pulp-toolchain.
  
The previous `.word` format was indeed a pain, but one can alternatively also spell these as:
```suggestion
    asm volatile(
        // dmsrc %[sl], %[sh]
        // opcode=0x2b, func3=0, func7=0, rd, rs1, rs2
        ".insn r 0x2b, 0, 0, zero, %[sl], %[sh]\n"
       ...
```
This should get you the same benefits in regards to giving the compiler the most flexibility for better register allocation (unlike the `.word`) while not tying the code more to the pulp-toolchain than it needs to.

Reply:
Hi @zero9178,
Unfortunately, we are still using an old LLVM 12.0 pulp toolchain, which does not support the `.insn` directive it seems. But we can keep the `.word` format in that case.